components/resource_control/src/worker.rs

```diff
@@ -0,0 +1,425 @@
+                    Entry::Occupied(mut s) => total_stats - s.insert(total_stats),
+                    Entry::Vacant(v) => {
+                        v.insert(total_stats);
+                        total_stats
+                    }
+                };
+            stats_delta = stats_delta / dur_secs;
+            background_consumed_total += stats_delta.total_consumed as f64;
+            g.stats = stats_delta;
+            if stats_delta.total_wait_dur_us > 0 {
+                has_wait = true;
+            }
+        }
+
+        // fast path if process cpu is low
+        let is_low_load = resource_stats.current_used <= (resource_stats.total_quota * 0.1);
+        if is_low_load && !has_wait && self.is_last_time_low_load[resource_type as usize] {
+            return;
+        }
+        self.is_last_time_low_load[resource_type as usize] = is_low_load;
+
+        let mut available_quota = ((resource_stats.total_quota - resource_stats.current_used
+            + background_consumed_total)
+            * 0.9)
+            .max(resource_stats.total_quota * 0.1);
+        let mut total_expected_cost = 0.0;
+        for g in bg_group_stats.iter_mut() {
+            let mut rate_limit = limiter_fn(&g.limiter).get_rate_limit();
+            if rate_limit.is_infinite() {
+                rate_limit = 0.0;
+            }
+            let group_expected_cost = g.stats.total_consumed as f64
+                + g.stats.total_wait_dur_us as f64 / 1000000.0 * rate_limit;
+            g.expect_cost_per_ru = group_expected_cost / g.ru_quota;
+            total_expected_cost += group_expected_cost;
+        }
+        bg_group_stats.sort_by(|g1, g2| {
+            g1.expect_cost_per_ru
+                .partial_cmp(&g2.expect_cost_per_ru)
+                .unwrap()
+        });
+
+        // quota is enough
+        if total_expected_cost <= available_quota {
+            for g in bg_group_stats.iter().rev() {
+                let expected = g.expect_cost_per_ru * g.ru_quota;
+                let limit = if g.expect_cost_per_ru > available_quota / total_ru_quota {
+                    expected
+                } else {
+                    available_quota / total_ru_quota * g.ru_quota
+                };
+                limiter_fn(&g.limiter).set_rate_limit(limit);
+                available_quota -= limit;
+                total_ru_quota -= g.ru_quota;
+            }
+            return;
+        }
+
+        // quota is not enough
+        for g in bg_group_stats {
+            let expected = g.expect_cost_per_ru * g.ru_quota;
+            let limit = if g.expect_cost_per_ru < available_quota / total_ru_quota {
+                expected
+            } else {
+                available_quota / total_ru_quota * g.ru_quota
+            };
+            limiter_fn(&g.limiter).set_rate_limit(limit);
+            available_quota -= limit;
+            total_ru_quota -= g.ru_quota;
+        }
```

Comment:
It looks like it will go to meet the first in arr, Do we need to divide it proportionally?

Reply:
I don't get what you mean. BTW the array is sorted in the order of required quota per ru